hashprobe/hashprobe.go

```diff
@@ -0,0 +1,643 @@
+// Package hashprobe provides implementations of probing tables for various
+// data types.
+//
+// Probing tables are specialized hash tables supporting only a single
+// "probing" operation which behave like a "lookup or insert". When a key
+// is probed, either its value is retrieved if it already existed in the table,
+// or it is inserted and assigned its index in the insert sequence as value.
+//
+// Values are represented as signed 32 bits integers, which means that probing
+// tables defined in this package may contain at most 2^31-1 entries.
+//
+// Probing tables have a method named Probe with the following signature:
+//
+//	func (t *Int64Table) Probe(keys []int64, values []int32) int {
+//		...
+//	}
+//
+// The method takes an array of keys to probe as first argument, an array of
+// values where the indexes of each key will be written as second argument, and
+// returns the number of keys that were inserted during the call.
+//
+// Applications that need to determine which keys were inserted can capture the
+// length of the probing table prior to the call, and scan the list of values
+// looking for indexes greater or equal to the captured lenth.
+package hashprobe
+
+import (
+	"math"
+	"math/bits"
+	"math/rand"
+
+	"github.com/segmentio/parquet-go/hashprobe/aeshash"
+	"github.com/segmentio/parquet-go/hashprobe/wyhash"
+	"github.com/segmentio/parquet-go/internal/unsafecast"
+)
+
+const (
+	// Number of probes tested per iteration. This parameter balances between
+	// the amount of memory allocated on the stack to hold the computed hashes
+	// of the keys being probed, and amortizing the baseline cost of the probing
+	// algorithm.
+	//
+	// The larger the value, the more memory is required, but lower the baseline
+	// cost will be.
+	//
+	// We chose a value that is somewhat large, resulting in reserving 2KiB of
+	// stack but mostly erasing the baseline cost.
+	probesPerLoop = 256
+)
+
+func nextPowerOf2(n int) int {
+	return 1 << (64 - bits.LeadingZeros64(uint64(n-1)))
+}
+
+func randSeed() uintptr {
+	return uintptr(rand.Uint64())
```

Comment:
Don't you need to seed rand first? You'll always get the same value otherwise

Reply:
I can improve this, I was thinking the application usually seeds `math/rand`, it would be weird for the parquet-go library to force seeding the global random number generator; the current form does seem like a potential landmine.